# Introduction

The mobile server works on various models to classify the spam and ham (not spam) sms.

These models do text recognition and on the basis of recognition, models predicts whether to store the sms in spam folder or not.We will also build a such type of model today and see how data science play such a wonderful role in our daily life by separating unwanted messages in spam.

### Problem
So our todays problem is to predicting spam mails through different Machine learning model.

### Solution 
First, We will download the dataset from <a href='https://www.kaggle.com/uciml/sms-spam-collection-dataset' target='_blank'>Kaggle </a>.Then we will do the feature engineering or extraction and make the data in suitable form to do model building.We will split the datset into train and test. Finally, We will create different Machine Learning model to solve our spam predictions problem and then check
the choose the best model by comparing each model.

### Load the Dataset
We will import the pandas to load the dataset the we have downloaded

In [1]:
# importing pandas
import pandas as pd

After importing the library, we will load the dataset using pandas attribute read_csv

`Make sure that you have downloaded the dataset in the same folder in which our script is otherwise it will create an error.`

In [2]:
# loading the dataset (encoding is done due to file format is latin)
df=pd.read_csv('spam.csv',encoding='latin-1')

### Analyzing
After loading the dataset we will do the analysis on the data to make it suitable for our model
Seeing the data by head attribute which will load the first 5 rows of our data 

In [3]:
# Taking a glance at the data
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

By seeing the data we can say that, the columns name is not in good format and there is other three rows namely 1.unnamed:2 2.unmaned:3 3.unnamed:4 which are having null values and not needed for analyzing so we will delete all the three rows one by one using del attribute 

In [4]:
# deleting unnamed:2
del df['Unnamed: 2']

In [5]:
# deleting unnamed:3
del df['Unnamed: 3']

In [6]:
# deleting unnamed:4
del df['Unnamed: 4']

After deleting the unwanted three columns we have to name our columns in specific format so that the dataset will be in good format. 
We will do this by changing the name of column v1 to Result and v2 to Text by rename attribute.

In [7]:
# changing columns name
df.rename(columns={'v1':'Result','v2':'Text'},inplace=True)

### Saving the dataset
After transforming the dataset into suitable format we can save the dataset to do further analysis.
`Note this step is optional but very important since we tranformed data in suitable form and now we can also share it to others if needed`

Saving the data using to_csv attribute of pandas

In [8]:
# saving the dataset (index=False is used to not creating another index during saving)
df.to_csv('spamham.csv',index=False)

After saving the dataset we dan load the new dataset through read_csv to do further analysis and model building

In [9]:
# reading the new dataset or loading the new dataset
df=pd.read_csv('spamham.csv')

After loading the new dataset we can print the first 5 rows of the dataset by using head

In [10]:
# taking a look at the dataset
df.head()

Result                                               Text
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

Now, the dataset is right format
Lets check the descriptive statistics of our data using describe attribute which will load the statistical analysis on the numerical dataset

In [11]:
# statistical analysis
df.describe()

Result                    Text
count    5572                    5572
unique      2                    5169
top       ham  Sorry, I'll call later
freq     4825                      30

There is two unique value in our dataset ham and spam in which ham having more in number i.e. 4825 keeping this in mind lets check the data type of the columns by using type attribute

In [12]:
# printing the data types
df.dtypes

Result    object
Text      object
dtype: object

Further we can tranform the result data type into numerical in which ham=0 and spam=1 to make the predicted more suitable

In [13]:
# passing result columns in a for-loop to change the ham and spam in 0 and 1 respectively
for result in df['Result']:
    if result=='ham':
        df['Result']=df['Result'].replace('ham',0)
    else:
        df['Result']=df['Result'].replace('spam',1)

Now we can look at the changed result column

In [14]:
# checking the result column
df['Result']

0       0
1       0
2       1
3       0
4       0
       ..
5567    1
5568    0
5569    0
5570    0
5571    0
Name: Result, Length: 5572, dtype: int64

In [15]:
# checking the changed data type of result
df.dtypes

Result     int64
Text      object
dtype: object

# Train-Test Split
Now, its time split the data into training and testing datset.

Its a very important step of our model building. We have to split the data into train and test so that the model will train from the training dataset and then predict or test the model using test data and train data
First, import the train test split from the sklearn library

In [16]:
# importing the train-test split package
from sklearn.model_selection import train_test_split

After importing the library we have to make X (independant variable) and Y (dependant variable) using our dataset

In [17]:
# initializing X and Y
X=df['Text']
Y=df['Result']

Now, we are good to split the data into train and test

Splitting the data in ratio of 80-20 of train-test 

In [18]:
# splitting the dataset
x_train,x_test,y_train,y_test=train_test_split(X,Y,train_size=0.8,test_size=0.2,random_state=3)

# Feature Extraction 
We cannot do the model building on the text data so we have the pull or extract the feature of the text on the basis of which our model will do the prediction 

We will import the feature extraction libarary in which we will use TfidfVectorizer to extracting the feature from our data 

In [19]:
# Importing feature extraction library
from sklearn.feature_extraction.text import TfidfVectorizer as vector

After importing feature_extraction, we will tranform our x_train and x_test data into x_train_features and x_test_features respectively in lowercase and initialise the stop words as english 

`Stop words means like various english letter which normally told us that whether the sms being arrived is ham or spam using words like "Congratulation,Hurray,you won" in ham"`

In [20]:
# initializing the feature extraction as feature
feature=vector(min_df=1,lowercase=True,stop_words='english')

Now its time to tranfrom the data using fit_tranfrom attribute of vector

In [21]:
#tranforming x_train
x_train_features=feature.fit_transform(x_train)

# tranforming x_test
x_test_features=feature.transform(x_test)

# Using Support Vector Machine
Finally, its time to load our model and build it then predict our model accuracy 

First, we will train our dataset using Support Vector Machine (SVM) model.Loading the model SVM using sklearn library
and also importing the accuracy_score to check the accuracy of our model

In [22]:
# importing SVM which is named as LinearSVC in sklearn
from sklearn.svm import LinearSVC

# importing accuracy score for checking accuracy
from sklearn.metrics import accuracy_score

Now,initializing the SVM to do further analysis

In [23]:
# initializing SVM as svm
svm=LinearSVC()

Then fit the training set into our model to train the model using fit attribute

In [24]:
# Training the svm
svm.fit(x_train_features,y_train)

LinearSVC()

When the output shows the original Library means our model is fully trained using SVM

Finally, Its time to predict or test the model using test data and by predict attribute and then checking the accuracy by comparing the true and predicted value using accuracy_score

In [25]:
# testing the model on train data
y_train_pred=svm.predict(x_train_features)

# checking its accuracy
acc_svm = accuracy_score(y_train,y_train_pred)
print("Accuracy score on train model :",acc_svm)

Accuracy score on train model : 0.9995512676688355


SO our model gives 99% accuracy on our train data 

Lets check the accuracy on the test data

In [26]:
# testing the model on test data
y_test_pred=svm.predict(x_test_features)
acc = accuracy_score(y_test,y_test_pred)
print("Accuracy score on train model :",acc)

Accuracy score on train model : 0.9856502242152466


So, our model gives 98% accuracy on our test data.
# Validating model
We can validate our model by giving it a new data so as manually the accuracy of our model on new data

In [27]:
# giving new input
input=['I HAVE A DATE ON SUNDAY WITH WILL!!']

But the input will not be given to our model as it is since it is in raw format and we have to do feature extraction also so we have to convert our data using the feature extraction we initialzed up and then check the outcome using predict function of our model 

In [28]:
# tranforming the input using fit attribute
input_feature=feature.transform(input)

# predicting the outcome
pred=svm.predict(input_feature)
if pred[0]==1:
    print('Spam Mail')
else:
    print('Ham Mail')

Ham Mail


SO, our model is going good on the new data.
# Using K-Nearest Neighbour
Now,lets check whether using KNearestNeighbour (KNN) will give us more accuracy or not

First, we have to import the KNN which is a package of sklearn and then as usual we have to train our model and then predict or test it and then check accuracy.After all this, we will compare the accuracy of the both model

In [29]:
# importing KNN
from sklearn.neighbors import KNeighborsClassifier

After importing we have to initialize the model with K=3

`This is done because our model will predict the outcome based on its neighbour`

In [30]:
# initialising the model
knn=KNeighborsClassifier(n_neighbors=3)

Training the model using fit attribute. The output will print the model means it is successfuly trained

In [31]:
# training our model
knn.fit(x_train_features,y_train)

KNeighborsClassifier(n_neighbors=3)

And finally we have to test our model and check its accuracy by comparing true and predicted value
 

In [32]:
# Testing the model on train data
y_train_pred=knn.predict(x_train_features)

# checking its accuracy
acc=accuracy_score(y_train,y_train_pred)
print("Accuracy on train score :",acc)

Accuracy on train score : 0.9450302894323536


So, our KNN model gives 94% accuracy on test data which is comparatively less than SVM model 

Lets check our accuracy on test data

In [33]:
# testing the model on test data
y_test_pred=knn.predict(x_test_features)

# checking its accuracy
acc_knn=accuracy_score(y_test,y_test_pred)
print("Accuracy on train score :",acc_knn)

Accuracy on train score : 0.9201793721973094


So, our KNN model gives 92% accuracy on test data which is also less than SVM model 
# Validating model
But it is not a bad model since it gives more than 90% accuracy on test and train data. SO we can also validate this model by using a new input as a validation data

In [34]:
# giving a new input
input=['I HAVE A DATE ON SUNDAY WITH WILL!!']

So we have to tranfrom this new input into suitable format using feature extraction which we had imported earlier

In [35]:
# transforming new data for passing into the model
input_feature=feature.transform(input)

# predicting the outcome
pred=knn.predict(input_feature)
if pred[0]==1:
    print('Spam Mail')
else:
    print('Ham Mail')

Ham Mail


# Conclusion
So our both model gives a good accuracy on new data but in comparison the SVM is leading 
In this project, We have successfully build our model using SVM and KNN model of sklearn and solved our problem

# Thank You
I want to thank you all for reading my post, I hope that now you can easily build a binary classification model using raw dataset. But if any problem occurs, contact me in comment section and any appreciation are also most welcome.
Thank You!!!